pkg/controllers/webhook/controller.go

```diff
@@ -634,74 +634,85 @@ func (c *controller) buildResourceMutatingWebhookConfiguration(ctx context.Conte
 		Webhooks:   []admissionregistrationv1.MutatingWebhook{},
 	}
 	if c.watchdogCheck() {
-		ignore := newWebhook(c.defaultTimeout, ignore)
-		fail := newWebhook(c.defaultTimeout, fail)
+		webhookCfg := config.WebhookConfig{}
+		webhookCfgs := cfg.GetWebhooks()
+		if len(webhookCfgs) > 0 {
+			webhookCfg = webhookCfgs[0]
+		}
+
+		ignoreWebhook := newWebhook(c.defaultTimeout, ignore, cfg.GetMatchConditions())
+		failWebhook := newWebhook(c.defaultTimeout, fail, cfg.GetMatchConditions())
 		policies, err := c.getAllPolicies()
 		if err != nil {
 			return nil, err
 		}
+		var fineGrainedIgnoreList, fineGrainedFailList []*webhook
 		c.recordPolicyState(config.MutatingWebhookConfigurationName, policies...)
 		for _, p := range policies {
 			if p.AdmissionProcessingEnabled() {
 				spec := p.GetSpec()
 				if spec.HasMutateStandard() || spec.HasVerifyImages() {
+					if spec.CustomWebhookConfiguration() {
+						fineGrainedIgnore := newWebhookPerPolicy(c.defaultTimeout, ignore, cfg.GetMatchConditions(), p)
+						fineGrainedFail := newWebhookPerPolicy(c.defaultTimeout, fail, cfg.GetMatchConditions(), p)
+						if spec.GetFailurePolicy(ctx) == kyvernov1.Ignore {
+							c.mergeWebhook(fineGrainedIgnore, p, false)
+							fineGrainedIgnoreList = append(fineGrainedIgnoreList, fineGrainedIgnore)
+						} else {
+							c.mergeWebhook(fineGrainedFail, p, false)
+							fineGrainedFailList = append(fineGrainedFailList, fineGrainedFail)
+						}
+						continue
+					}
+
 					if spec.GetFailurePolicy(ctx) == kyvernov1.Ignore {
-						c.mergeWebhook(ignore, p, false)
+						c.mergeWebhook(ignoreWebhook, p, false)
 					} else {
-						c.mergeWebhook(fail, p, false)
+						c.mergeWebhook(failWebhook, p, false)
 					}
 				}
 			}
 		}
-		webhookCfg := config.WebhookConfig{}
-		webhookCfgs := cfg.GetWebhooks()
-		if len(webhookCfgs) > 0 {
-			webhookCfg = webhookCfgs[0]
-		}
```

Comment:
@realshuting - It seems that this code already exists before this PR. I don't think it is related to policy's webhook configuration

Reply:
ok